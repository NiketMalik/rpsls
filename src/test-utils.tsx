import { FC, ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { THEME } from "constants/theme";

import { StoreProvider } from "context/storeContext";
import { ThemeProvider } from "@emotion/react";

const WithGlobalProviders: FC = ({ children }) => {
  return (
    <StoreProvider>
      <ThemeProvider theme={THEME}>{children}</ThemeProvider>
    </StoreProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithGlobalProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
