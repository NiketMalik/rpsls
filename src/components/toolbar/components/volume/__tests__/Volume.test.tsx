import React from "react";
import ReactDOM from "react-dom";
import { fireEvent, render, screen, waitFor, act } from "test-utils";

import { Volume } from "../index";

const mockAudioPlay = jest.fn();
const mockAudioStop = jest.fn();

jest.mock("hooks/useAudio", () => ({
  useAudio: () => [mockAudioPlay, { stop: mockAudioStop }],
}));

describe("Volume", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders without error", () => {
    const container = document.createElement("div");
    act(() => {
      ReactDOM.render(<Volume />, container);
    });

    expect(container).toMatchSnapshot();
  });

  describe("tooltip actions", () => {
    it("should show tooltip for audio", async () => {
      render(<Volume />);

      const audioTriggerEle = screen.getByTestId("audio-trigger");
      expect(audioTriggerEle).toBeInTheDocument();

      fireEvent.mouseOver(audioTriggerEle);

      await waitFor(() => {
        const triggerTooltipEle = screen.getByTestId("audio-tooltip");
        expect(triggerTooltipEle).toBeInTheDocument();
        expect(triggerTooltipEle).toMatchSnapshot();
      });
    });

    it("should show tooltip for music", async () => {
      render(<Volume />);

      const musicTriggerEle = screen.getByTestId("music-trigger");
      expect(musicTriggerEle).toBeInTheDocument();

      fireEvent.mouseOver(musicTriggerEle);

      await waitFor(() => {
        const triggerTooltipEle = screen.getByTestId("music-tooltip");
        expect(triggerTooltipEle).toBeInTheDocument();
        expect(triggerTooltipEle).toMatchSnapshot();
      });
    });
  });

  describe("button actions", () => {
    it("should enable audio", async () => {
      render(<Volume />);

      const audioTriggerEle = screen.getByTestId("audio-trigger");
      expect(audioTriggerEle).toBeInTheDocument();

      fireEvent.click(audioTriggerEle);
      fireEvent.mouseOver(audioTriggerEle);

      await waitFor(() => {
        const triggerTooltipEle = screen.getByTestId("audio-tooltip");
        expect(triggerTooltipEle).toBeInTheDocument();
      });

      await waitFor(() => expect(mockAudioPlay).toBeCalled());
    });

    it("should disable audio", async () => {
      render(<Volume />);

      const audioTriggerEle = screen.getByTestId("audio-trigger");
      expect(audioTriggerEle).toBeInTheDocument();

      fireEvent.mouseOver(audioTriggerEle);
      await waitFor(() => {
        const triggerTooltipEle = screen.getByTestId("audio-tooltip");
        expect(triggerTooltipEle).toBeInTheDocument();
        expect(triggerTooltipEle).toMatchSnapshot();
      });

      fireEvent.click(audioTriggerEle);
      await waitFor(() => expect(mockAudioPlay).toBeCalledTimes(1));

      fireEvent.mouseOver(audioTriggerEle);
      await waitFor(() => {
        const triggerTooltipEle = screen.getByTestId("audio-tooltip");
        expect(triggerTooltipEle).toBeInTheDocument();
        expect(triggerTooltipEle).toMatchSnapshot();
      });
    });

    it("should enable music", async () => {
      render(<Volume />);

      const musicTriggerEle = screen.getByTestId("music-trigger");
      expect(musicTriggerEle).toBeInTheDocument();

      fireEvent.click(musicTriggerEle);
      fireEvent.mouseOver(musicTriggerEle);

      await waitFor(() => {
        const triggerTooltipEle = screen.getByTestId("music-tooltip");
        expect(triggerTooltipEle).toBeInTheDocument();
      });

      await waitFor(() => expect(mockAudioPlay).not.toBeCalled());
    });
  });
});
