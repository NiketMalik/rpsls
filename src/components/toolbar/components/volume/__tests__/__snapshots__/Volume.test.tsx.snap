// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Volume button actions should disable audio 1`] = `
<div
  class="fade show tooltip bs-tooltip-top"
  data-testid="audio-tooltip"
  id="toolbar-audio-tooltip"
  role="tooltip"
  x-placement="top"
>
  <div
    class="tooltip-arrow"
  />
  <div
    class="tooltip-inner"
  >
    Enable
     game audio
  </div>
</div>
`;

exports[`Volume button actions should disable audio 2`] = `
<div
  class="fade show tooltip bs-tooltip-top"
  data-popper-escaped="true"
  data-popper-placement="top"
  data-popper-reference-hidden="true"
  data-testid="audio-tooltip"
  id="toolbar-audio-tooltip"
  role="tooltip"
  style="position: absolute; bottom: 0px; left: 0px; transform: translate(0px, 0px);"
  x-placement="top"
>
  <div
    class="tooltip-arrow"
    style="position: absolute; left: 0px; transform: translate(0px, 0px);"
  />
  <div
    class="tooltip-inner"
  >
    Enable
     game audio
  </div>
</div>
`;

exports[`Volume renders without error 1`] = `
<div>
  <button
    aria-label="audio"
    class="btn btn-transparent"
    data-testid="audio-trigger"
    type="button"
  >
    <img
      alt="audio volume"
      class=""
      height="32"
      src="icon-volume-mute.svg"
      width="32"
    />
  </button>
  <button
    aria-label="music"
    class="btn btn-transparent"
    data-testid="music-trigger"
    type="button"
  >
    <img
      alt="music volume"
      class=""
      height="28"
      src="icon-audio-mute.svg"
      width="28"
    />
  </button>
</div>
`;

exports[`Volume tooltip actions should show tooltip for audio 1`] = `
<div
  class="fade show tooltip bs-tooltip-top"
  data-testid="audio-tooltip"
  id="toolbar-audio-tooltip"
  role="tooltip"
  x-placement="top"
>
  <div
    class="tooltip-arrow"
  />
  <div
    class="tooltip-inner"
  >
    Enable
     game audio
  </div>
</div>
`;

exports[`Volume tooltip actions should show tooltip for music 1`] = `
<div
  class="fade show tooltip bs-tooltip-top"
  data-testid="music-tooltip"
  id="toolbar-music-tooltip"
  role="tooltip"
  x-placement="top"
>
  <div
    class="tooltip-arrow"
  />
  <div
    class="tooltip-inner"
  >
    Enable
     game music
  </div>
</div>
`;
